### ISSUE TITLE
Various html5lib improvements

### ISSUE THREAD

**Issue by kovidgoyal** [html5lib/html5lib-python#119]
This is just a FYI, I have been working on a modified version of html5lib that achieves the following goals:

1) Preserves attribute order
2) Optionally includes line and column number information when parsing
3) Handles XML namespaces correctly, so that if you happen to parse an XHTML document with html5lib you dont lose all the namespace information
4) Create a new treebuilder for lxml
5) Various performance improvements

Using my new lxml treebuilder parsing performance with line numbers and attribute order preservation is the same as for vanilla html5lib with its builtin treebuilder. The speed improvements come mainly from the new lxml builder and an optimized inputstream class for in memory streams.

I make no claims as to the relevance of my work for html5lib. I am just sharing it with you as a way of giving back. You are welcome to use the patches or not. Feel free to ask if you need any clarification.

The code is in https://github.com/kovidgoyal/calibre/tree/master/src/html5lib (these are the changes to html5lib itself)

and the lxml builder is in:
https://github.com/kovidgoyal/calibre/blob/master/src/calibre/ebooks/oeb/polish/parsing.py

**Comment by scoder** [comment]
> Element.attrib (which generates a python dictionary from the underlying libxml2 structure).

No, it actually uses its own mapping class based on the underlying XML tree data. The order is therefore preserved.

> lxml, FWIW, explicitly documents everywhere that the order is "arbitrary".

"Arbitrary" in the sense that it depends on where the attributes came from. If from the parser, then it's document order. If added later, it might make a difference whether the attribute already existed before, but AFAIR it still gets appended at the end. It's up to libxml2, and lxml doesn't give any guarantees here on its own. It's unlikely that the current behaviour of libxml2 is ever going to change, though, as it is considered a feature.
